File: test/php/disbursement/DanaAccountInquiryTest.php
<?php

namespace DanaUat\Disbursement;

use PHPUnit\Framework\TestCase;
use Dana\Disbursement\v1\Api\v1\Api\DisbursementApi;
use Dana\ObjectSerializer;
use Dana\Configuration;
use Dana\Env;
use Dana\ApiException;
use Dana\Disbursement\v1\Api\DisbursementApi as ApiDisbursementApi;
use DanaUat\Helper\Assertion;
use DanaUat\Helper\Util;
use Exception;

class DanaAccountInquiryTest extends TestCase
{
    private static $titleCase = 'DanaAccountInquiry';
    private static $jsonPathFile = 'resource/request/components/Disbursement.json';
    private static $apiInstance;
    private static $merchantId;

    public static function setUpBeforeClass(): void
    {
        // Set up configuration with authentication settings
        $configuration = new Configuration();

        // The Configuration constructor automatically loads values from environment variables
        // But we can manually set them if needed
        $configuration->setApiKey('PRIVATE_KEY', getenv('PRIVATE_KEY'));
        $configuration->setApiKey('ORIGIN', getenv('ORIGIN'));
        $configuration->setApiKey('X_PARTNER_ID', getenv('X_PARTNER_ID'));
        $configuration->setApiKey('ENV', Env::SANDBOX);

        // Create DisbursementApi instance directly with configuration
        self::$apiInstance = new ApiDisbursementApi(null, $configuration);

        // Store merchantId for reuse
        self::$merchantId = getenv('MERCHANT_ID');
    }

    /**
     * Should successfully inquire DANA account with valid data
     */
    public function testInquiryCustomerValidData(): void
    {
        Util::withDelay(function () {
            $caseName = 'InquiryCustomerValidData';

            try {
                // Get and prepare the request
                $jsonDict = Util::getRequest(
                    self::$jsonPathFile,
                    self::$titleCase,
                    $caseName
                );

                // Set a unique partner reference number
                $partnerReferenceNo = Util::generatePartnerReferenceNo();
                $jsonDict['partnerReferenceNo'] = $partnerReferenceNo;

                // Create a DanaAccountInquiryRequest object from the JSON request data
                $danaAccountInquiryRequestObj = ObjectSerializer::deserialize(
                    $jsonDict,
                    'Dana\Disbursement\v1\Model\DanaAccountInquiryRequest'
                );

                // Make the API call
                $apiResponse = self::$apiInstance->danaAccountInquiry($danaAccountInquiryRequestObj);

                // Assert the response matches the expected data
                Assertion::assertResponse(
                    self::$jsonPathFile,
                    self::$titleCase,
                    $caseName,
                    $apiResponse->__toString(),
                    ['partnerReferenceNo' => $partnerReferenceNo]
                );

                $this->assertTrue(true);
            } catch (ApiException $e) {
                $this->fail('Failed to inquire DANA account: ' . $e->getMessage());
            } catch (Exception $e) {
                $this->fail('Unexpected exception: ' . $e->getMessage());
            }
        });
    }

    /**
     * Should fail inquiry due to unauthorized signature
     */
    public function testInquiryCustomerUnauthorizedSignature(): void
    {
        try {
            Util::withDelay(function() {
                $caseName = 'InquiryCustomerUnauthorizedSignature';
                $partnerReferenceNo = Util::generatePartnerReferenceNo();
                
                // Get the request data from the JSON file
                $requestData = Util::getRequest(
                    self::$jsonPathFile, 
                    self::$titleCase, 
                    $caseName
                );
                
                // Set the correct partner reference number
                $requestData['partnerReferenceNo'] = $partnerReferenceNo;
                
                // Create headers with invalid signature to test authorization failure
                $headers = Util::getHeadersWithSignature(
                    'POST', 
                    '/rest/v1.0/emoney/account-inquiry',
                    $requestData,
                    true,
                    false,
                    true  // invalid_signature set to true
                );
                
                // Make direct API call with invalid signature
                try {
                    Util::executeApiRequest(
                        'POST',
                        'https://api.sandbox.dana.id/rest/v1.0/emoney/account-inquiry',
                        $headers,
                        $requestData
                    );
                    
                    $this->fail('Expected ApiException for invalid signature but the API call succeeded');
                } catch (ApiException $e) {
                    // We expect a 401 Unauthorized for invalid signature
                    $this->assertEquals(401, $e->getCode(), "Expected HTTP 401 Unauthorized for invalid signature, got {$e->getCode()}");

                    // Get the response body from the exception
                    $responseContent = (string)$e->getResponseBody();
                    
                    // Use assertFailResponse to validate the error response
                    Assertion::assertFailResponse(
                        self::$jsonPathFile, 
                        self::$titleCase, 
                        $caseName, 
                        $responseContent,
                        ['partnerReferenceNo' => $partnerReferenceNo]
                    );
                } catch (Exception $e) {
                    $this->fail("Expected ApiException but got " . get_class($e) . ": " . $e->getMessage());
                }
            });
        } catch (Exception $e) {            
            $this->fail("Unexpected exception: " . $e->getMessage());
        }
    }

    /**
     * Should fail inquiry due to frozen account
     */
    public function testInquiryCustomerFrozenAccount(): void
    {
        Util::withDelay(function () {
            $caseName = 'InquiryCustomerFrozenAccount';

            // Get and prepare the request
            $jsonDict = Util::getRequest(
                self::$jsonPathFile,
                self::$titleCase,
                $caseName
            );

            // Set a unique partner reference number
            $partnerReferenceNo = Util::generatePartnerReferenceNo();
            $jsonDict['partnerReferenceNo'] = $partnerReferenceNo;

            // Create a DanaAccountInquiryRequest object from the JSON request data
            $danaAccountInquiryRequestObj = ObjectSerializer::deserialize(
                $jsonDict,
                'Dana\Disbursement\v1\Model\DanaAccountInquiryRequest'
            );

            try {
                // Make the API call
                self::$apiInstance->danaAccountInquiry($danaAccountInquiryRequestObj);

                $this->fail('Expected ApiException for frozen account but the API call succeeded');
            } catch (ApiException $e) {
                // We expect a 403 Forbidden for frozen account
                $this->assertEquals(403, $e->getCode(), "Expected HTTP 403 Forbidden for frozen account, got {$e->getCode()}");

                // Get the response body from the exception
                $responseContent = (string)$e->getResponseBody();
                
                // Use assertFailResponse to validate the error response
                Assertion::assertFailResponse(
                    self::$jsonPathFile, 
                    self::$titleCase, 
                    $caseName, 
                    $responseContent,
                    ['partnerReferenceNo' => $partnerReferenceNo]
                );
            } catch (Exception $e) {
                $this->fail("Expected ApiException but got " . get_class($e) . ": " . $e->getMessage());
            }
        });
    }

    /**
     * Should fail inquiry due to unregistered account
     */
    public function testInquiryCustomerUnregisteredAccount(): void
    {
        Util::withDelay(function () {
            $caseName = 'InquiryCustomerUnregisteredAccount';

            // Get and prepare the request
            $jsonDict = Util::getRequest(
                self::$jsonPathFile,
                self::$titleCase,
                $caseName
            );

            // Set a unique partner reference number
            $partnerReferenceNo = Util::generatePartnerReferenceNo();
            $jsonDict['partnerReferenceNo'] = $partnerReferenceNo;

            // Create a DanaAccountInquiryRequest object from the JSON request data
            $danaAccountInquiryRequestObj = ObjectSerializer::deserialize(
                $jsonDict,
                'Dana\Disbursement\v1\Model\DanaAccountInquiryRequest'
            );

            try {
                // Make the API call
                self::$apiInstance->danaAccountInquiry($danaAccountInquiryRequestObj);

                $this->fail('Expected ApiException for unregistered account but the API call succeeded');
            } catch (ApiException $e) {
                // We expect a 403 Forbidden for unregistered account
                $this->assertEquals(403, $e->getCode(), "Expected HTTP 403 Forbidden for unregistered account, got {$e->getCode()}");

                // Get the response body from the exception
                $responseContent = (string)$e->getResponseBody();
                
                // Use assertFailResponse to validate the error response
                Assertion::assertFailResponse(
                    self::$jsonPathFile, 
                    self::$titleCase, 
                    $caseName, 
                    $responseContent,
                    ['partnerReferenceNo' => $partnerReferenceNo]
                );
            } catch (Exception $e) {
                $this->fail("Expected ApiException but got " . get_class($e) . ": " . $e->getMessage());
            }
        });
    }

    /**
     * Should fail inquiry due to exceeded limit
     */
    public function testInquiryCustomerExceededLimit(): void
    {
        Util::withDelay(function () {
            $caseName = 'InquiryCustomerExceededLimit';

            // Get and prepare the request
            $jsonDict = Util::getRequest(
                self::$jsonPathFile,
                self::$titleCase,
                $caseName
            );

            // Set a unique partner reference number
            $partnerReferenceNo = Util::generatePartnerReferenceNo();
            $jsonDict['partnerReferenceNo'] = $partnerReferenceNo;

            // Create a DanaAccountInquiryRequest object from the JSON request data
            $danaAccountInquiryRequestObj = ObjectSerializer::deserialize(
                $jsonDict,
                'Dana\Disbursement\v1\Model\DanaAccountInquiryRequest'
            );

            try {
                // Make the API call
                self::$apiInstance->danaAccountInquiry($danaAccountInquiryRequestObj);

                $this->fail('Expected ApiException for exceeded limit but the API call succeeded');
            } catch (ApiException $e) {
                // We expect a 403 Forbidden for exceeded limit
                $this->assertEquals(403, $e->getCode(), "Expected HTTP 403 Forbidden for exceeded limit, got {$e->getCode()}");

                // Get the response body from the exception
                $responseContent = (string)$e->getResponseBody();
                
                // Use assertFailResponse to validate the error response
                Assertion::assertFailResponse(
                    self::$jsonPathFile, 
                    self::$titleCase, 
                    $caseName, 
                    $responseContent,
                    ['partnerReferenceNo' => $partnerReferenceNo]
                );
            } catch (Exception $e) {
                $this->fail("Expected ApiException but got " . get_class($e) . ": " . $e->getMessage());
            }
        });
    }
}

File: test/python/disbursement/dana_account_inquiry_test.py
import os
import pytest
from dana.utils.snap_configuration import SnapConfiguration, AuthSettings, Env
from dana.disbursement.v1.enum import *
from dana.disbursement.v1.models import *
from dana.disbursement.v1 import *
from dana.disbursement.v1.api import *
from dana.api_client import ApiClient
from dana.exceptions import *
from uuid import uuid4
from helper.util import get_request, with_delay
from helper.api_helpers import get_headers_with_signature, execute_and_assert_api_error
from helper.assertion import assert_response, assert_fail_response

title_case = "DanaAccountInquiry"
json_path_file = "resource/request/components/Disbursement.json"

configuration = SnapConfiguration(
    api_key=AuthSettings(
        PRIVATE_KEY=os.environ.get("PRIVATE_KEY"),
        ORIGIN=os.environ.get("ORIGIN"),
        X_PARTNER_ID=os.environ.get("X_PARTNER_ID"),
        DANA_ENV=Env.SANDBOX,
        X_DEBUG="true",
        CLIENT_SECRET=os.environ.get("CLIENT_SECRET"),
    )
)

with ApiClient(configuration) as api_client:
    api_instance = DisbursementApi(api_client)

@with_delay()
def test_inquiry_customer_valid_data():
    case_name = "InquiryCustomerValidData"
    json_dict = get_request(json_path_file, title_case, case_name)
    
    json_dict["partnerReferenceNo"] = str(uuid4())
    request_obj = DanaAccountInquiryRequest.from_dict(json_dict)

    try:
        response = api_instance.dana_account_inquiry(request_obj)
        assert_response(json_path_file, title_case, case_name, DanaAccountInquiryResponse.to_json(response))

    except ApiException as e:
        pytest.fail(f"API call failed: {e}")

@with_delay()
def test_inquiry_customer_unauthorized_signature():
    case_name = "InquiryCustomerUnauthorizedSignature"
    json_dict = get_request(json_path_file, title_case, case_name)
    
    json_dict["partnerReferenceNo"] = str(uuid4())
    request_obj = DanaAccountInquiryRequest.from_dict(json_dict)

    headers = get_headers_with_signature(
        method="POST",
        resource_path="v1.0/emoney/account-inquiry.htm",
        request_obj=json_dict,
        with_timestamp=True,
        invalid_timestamp=False
    )
    
    headers["X-SIGNATURE"] = "85be817c55b2c135157c7e89f52499bf0c25ad6eeebe04a986e8c862561b19a5"

    execute_and_assert_api_error(
        api_client,
        "POST",
        "https://api.sandbox.dana.id/rest/v1.0/emoney/account-inquiry",
        request_obj,
        headers,
        401,
        json_path_file,
        title_case,
        case_name
    )

@with_delay()
def test_inquiry_customer_frozen_account():
    case_name = "InquiryCustomerFrozenAccount"
    json_dict = get_request(json_path_file, title_case, case_name)
    
    json_dict["partnerReferenceNo"] = str(uuid4())
    request_obj = DanaAccountInquiryRequest.from_dict(json_dict)

    try:
        api_instance.dana_account_inquiry(request_obj)
        pytest.fail("Expected ApiException for frozen account but the API call succeeded")

    except ApiException as e:
        assert_fail_response(json_path_file, title_case, case_name, str(e.body), 
                           {'partnerReferenceNo': json_dict["partnerReferenceNo"]})

@with_delay()
def test_inquiry_customer_unregistered_account():
    case_name = "InquiryCustomerUnregisteredAccount"
    json_dict = get_request(json_path_file, title_case, case_name)
    
    json_dict["partnerReferenceNo"] = str(uuid4())
    request_obj = DanaAccountInquiryRequest.from_dict(json_dict)

    try:
        api_instance.dana_account_inquiry(request_obj)
        pytest.fail("Expected ApiException for unregistered account but the API call succeeded")

    except ApiException as e:
        assert_fail_response(json_path_file, title_case, case_name, str(e.body), 
                           {'partnerReferenceNo': json_dict["partnerReferenceNo"]})

@with_delay()
def test_inquiry_customer_exceeded_limit():
    case_name = "InquiryCustomerExceededLimit"
    json_dict = get_request(json_path_file, title_case, case_name)
    
    json_dict["partnerReferenceNo"] = str(uuid4())
    request_obj = DanaAccountInquiryRequest.from_dict(json_dict)

    try:
        api_instance.dana_account_inquiry(request_obj)
        pytest.fail("Expected ApiException for exceeded limit but the API call succeeded")

    except ApiException as e:
        assert_fail_response(json_path_file, title_case, case_name, str(e.body), 
                           {'partnerReferenceNo': json_dict["partnerReferenceNo"]})
